ArjunaJTA/jta/tests/classes/com/hp/mwtests/ts/jta/recovery/XARecoveryModuleUnitTest.java

```diff
@@ -374,6 +378,106 @@ public void start(Xid xid, int i) throws XAException {
+        XARecoveryModule xarm = new XARecoveryModule();
+        xarm.addXAResourceOrphanFilter(new JTANodeNameXAResourceOrphanFilter());
+        xarm.addXAResourceRecoveryHelper(new XAResourceRecoveryHelper() {
+            XAResource[] xares = new XAResource[]{new XAResource() {
+                int count = 0;
+                Xid xid = new XidImple(new Uid());
+
+                @Override
+                public void commit(Xid xid, boolean b) throws XAException {
+
+                }
+
+                @Override
+                public void end(Xid xid, int i) throws XAException {
+
+                }
+
+                @Override
+                public void forget(Xid xid) throws XAException {
+
+                }
+
+                @Override
+                public int getTransactionTimeout() throws XAException {
+                    return 0;
+                }
+
+                @Override
+                public boolean isSameRM(XAResource xaResource) throws XAException {
+                    return false;
+                }
+
+                @Override
+                public int prepare(Xid xid) throws XAException {
+                    return 0;
+                }
+
+                @Override
+                public Xid[] recover(int i) throws XAException {
+                    count++;
+                    if (count == 1 || count == 5) {
+                        return new Xid[]{xid};
+                    } else if (count > 5) {
+                        return new Xid[0];
+                    } else {
+                        throw new XAException();
+                    }
+                }
+
+                @Override
+                public void rollback(Xid xid) throws XAException {
+                    if (count == 1) { // This comes from the first end scan
+                        throw new XAException(XAException.XA_RETRY);
+                    }
+                    rolledback = true;
+                }
+
+                @Override
+                public boolean setTransactionTimeout(int i) throws XAException {
+                    return false;
+                }
+
+                @Override
+                public void start(Xid xid, int i) throws XAException {
+
+                }
+            }
+            };
+
+            @Override
+            public boolean initialise(String p) throws Exception {
+                return false;
+            }
+
+            @Override
+            public XAResource[] getXAResources() throws Exception {
+                return xares;
+            }
+        });
+
+        xarm.periodicWorkFirstPass();
+        xarm.periodicWorkSecondPass();
+        assertFalse(rolledback);
+        xarm.periodicWorkFirstPass();
+        xarm.periodicWorkSecondPass();
+        assertFalse(rolledback);
+        xarm.periodicWorkFirstPass();
+        xarm.periodicWorkSecondPass();
+        assertTrue(rolledback);
```

Comment:
A comment would be useful here (since the logic requires some insider knowledge), something like

> * 1st pass: returns one xid (count is 1)
> * 2nd pass: throws an exception (count is 2)
> * 1st pass: throws an exception (count is 3)
> * 2nd pass: throws an exception (count is 4)
> * 1st pass: returns an empty list of xids (count is 5)
> * 2nd pass: returns an empty list of xids (count is 6)

> So the first two recovery cycles do nothing with the resource (because phase two is getting the exception)
> When count reaches 6 it sees that the xid has gone and presumes abort so calls rollback and hence assertTrue(rolledback) passes